src/state_circuit/state.rs

```diff
@@ -0,0 +1,896 @@
+
+use std::marker::PhantomData;
+
+/*
+Example memory table:
+| address | global_counter | value | flag |
+---------------------------------
+|    0    |   0  |   0   |   1  |
+|    0    |  12  |  12   |   1  |
+|    0    |  24  |  12   |   0  |
+|    1    |   0  |   0   |   1  |
+|    1    |  17  |  32   |   1  |
+|    1    |  89  |  32   |   0  |
+*/
+
+/*
+Example bus mapping:
+
+| global_counter | memory_flag | memory_address | memory_value |
+------------------------------------------------------
+|  12  |      1      |        0       |      12      |
+|  17  |      1      |        1       |      32      |
+|  24  |      0      |        0       |      12      |
+|  89  |      0      |        1       |      32      |
+*/
+
+/// In the state proof, memory operations are ordered first by address, and then by global_counter.
+/// Memory is initialised at 0 for each new address.
+/// Memory is a word-addressed byte array, i.e. a mapping from a 253-bit word -> u8.
+#[derive(Copy, Clone, Debug)]
+struct Address<F: FieldExt>(F);
+
+/// Global counter
+#[derive(Copy, Clone, Debug)]
+struct GlobalCounter(usize);
+
+/// TODO: In the EVM we can only read memory values in 32 bytes, but can write either
+/// single-byte or 32-byte chunks. In zkEVM:
+/// - Are we reading single bytes or 32-byte chunks? TBD
+/// - We are writing a single field element (253 bits)
+#[derive(Copy, Clone, Debug)]
+struct Value<F: FieldExt>(F);
+
+#[derive(Clone, Debug)]
+enum ReadWrite<F: FieldExt> {
+    // flag == 0
+    Read(GlobalCounter, Value<F>),
+    // flag == 1
+    Write(GlobalCounter, Value<F>),
+}
+
+impl<F: FieldExt> ReadWrite<F> {
+    fn global_counter(&self) -> GlobalCounter {
+        match self {
+            Self::Read(global_counter, _) | Self::Write(global_counter, _) => *global_counter,
+        }
+    }
+
+    fn value(&self) -> Value<F> {
+        match self {
+            Self::Read(_, value) | Self::Write(_, value) => *value,
+        }
+    }
+
+    fn flag(&self) -> bool {
+        match self {
+            Self::Read(..) => false,
+            Self::Write(..) => true,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+/// All the read/write operations that happen at this address.
+pub(crate) struct Op<F: FieldExt> {
+    address: Address<F>,
+    global_counters: Vec<Option<ReadWrite<F>>>,
+}
+
+/// A mapping derived from witnessed memory operations.
+/// TODO: The complete version of this mapping will involve storage, stack,
+/// and opcode details as well.
+#[derive(Clone, Debug)]
+pub(crate) struct BusMapping<F: FieldExt> {
+    global_counter: Variable<usize, F>,
+    memory_flag: Variable<bool, F>,
+    memory_address: Variable<F, F>,
+    memory_value: Variable<F, F>,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct Config<
+    F: FieldExt,
+    const MAX_ROWS: usize,
+    const GLOBAL_COUNTER_MAX: usize,
+    const ADDRESS_MAX: usize,
+    const ADDRESS_INCR: bool,
+    const ADDRESS_DIFF_STRICT: bool,
```

Comment:
`ADDRESS_DIFF_STRICT` seems to always be `false` in the examples.
1. Is this correct -- do we never want to check strict monotonicity?
2. If it is always `false`, we could remove it from the const generics.

Reply:
Strict monotonicity is needed for global counter when the address stays the same, but there it's a bit tricky to use the monotone gadget and we don't use it currently (https://github.com/appliedzkp/zkevm-circuits/pull/23#discussion_r676472311). Agree, we can remove it, but I am just changing this in #24. Can I do it there?

Reply:
Removed now in this PR.